packages/material-ui/src/TextField/TextField.js

```diff
@@ -37,98 +42,176 @@ import Select from '../Select';
-    InputProps,
-    inputRef,
-    label,
-    multiline,
-    name,
-    onBlur,
-    onChange,
-    onFocus,
-    placeholder,
-    required,
-    rows,
-    rowsMax,
-    select,
-    SelectProps,
-    type,
-    value,
-    ...other
-  } = props;
+class TextField extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      width: 0,
+      height: 0,
+    };
+
+    this.inputRef = React.createRef();
+    this.labelRef = React.createRef();
+
+    this.resizeObserver = new ResizeObserver(this.handleResize);
+  }
+
+  componentDidMount() {
+    this.inputNode = ReactDOM.findDOMNode(this.inputRef.current);
+    this.labelNode = ReactDOM.findDOMNode(this.labelRef.current);
+
+    if (this.inputNode) {
+      this.resizeObserver.observe(this.inputNode);
+    }
+  }
+
+  componentWillUnmount() {
+    this.resizeObserver.disconnect();
+  }
+
+  get outlined() {
+    return this.props.variant === 'outlined';
+  }
+
+  get observeResize() {
+    return this.outlined;
+  }
+
+  get notchWidth() {
+    if (!this.labelNode) {
+      return 0;
+    }
+
+    return this.labelNode.offsetWidth * LABEL_SHRINK_SCALE + 8;
+  }
 
-  warning(
-    !select || Boolean(children),
-    'Material-UI: `children` must be passed when using the `TextField` component with `select`.',
-  );
+  get outline() {
+    if (!this.outlined) {
+      return null;
+    }
 
-  const helperTextId = helperText && id ? `${id}-helper-text` : undefined;
-  const InputElement = (
-    <Input
-      autoComplete={autoComplete}
-      autoFocus={autoFocus}
-      defaultValue={defaultValue}
-      disabled={disabled}
-      fullWidth={fullWidth}
-      multiline={multiline}
-      name={name}
-      rows={rows}
-      rowsMax={rowsMax}
-      type={type}
-      value={value}
-      id={id}
-      inputRef={inputRef}
-      onBlur={onBlur}
-      onChange={onChange}
-      onFocus={onFocus}
-      placeholder={placeholder}
-      inputProps={inputProps}
-      {...InputProps}
-    />
-  );
+    const { shrink } = this.props.InputLabelProps || {};
 
-  return (
-    <FormControl
-      aria-describedby={helperTextId}
-      className={className}
-      error={error}
-      fullWidth={fullWidth}
-      required={required}
-      {...other}
-    >
-      {label && (
-        <InputLabel htmlFor={id} {...InputLabelProps}>
-          {label}
-        </InputLabel>
-      )}
-      {select ? (
-        <Select value={value} input={InputElement} {...SelectProps}>
-          {children}
-        </Select>
-      ) : (
-        InputElement
-      )}
-      {helperText && (
-        <FormHelperText id={helperTextId} {...FormHelperTextProps}>
-          {helperText}
-        </FormHelperText>
-      )}
-    </FormControl>
-  );
+    return (
+      <NotchedOutline
+        select={this.props.select}
+        width={this.state.width}
+        height={this.state.height}
+        notched={shrink}
+        notchWidth={this.notchWidth}
+      />
+    );
+  }
+
```

Comment:
As with Input, these getters can be assigned to consts / inlined (as appropriate) in the render method, for example:

```jsx
    const notchWidth = this.labelNode ? this.labelNode.offsetWidth * 0.75 + 8 : 0;
    const outline =
      variant === 'outlined' ? (
        <NotchedOutline
          select={this.props.select}
          width={this.state.width}
          height={this.state.height}
          notched={InputLabelProps ? InputLabelProps.shrink : false}
          notchWidth={notchWidth}
        />
      ) : null;
```